tools/objinfo.ml

```diff
@@ -258,74 +259,135 @@ let read_dyn_header filename ic =
-    let crc = Digest.input ic in
-    close_in ic;
-    print_cmx_infos (ui, crc)
-  end else if magic_number = cmxa_magic_number then begin
-    let li = (input_value ic : library_infos) in
-    close_in ic;
-    print_cmxa_infos li
-  end else begin
-    let pos_trailer = in_channel_length ic - len_magic_number in
+  let open Magic_number in
+  let dump_standard ic =
+    match read_current_info ~expected_kind:None ic with
+      | Error ((Unexpected_error _) as err) ->
+         exit_magic_error ~expected_kind:None err
+      | Ok { kind; version = _ } ->
+         dump_obj_by_kind filename ic kind;
+         Ok ()
+      | Error (Parse_error head_error) ->
+         Error head_error
+  and dump_exec ic =
+    let pos_trailer = in_channel_length ic - Magic_number.magic_length in
     let _ = seek_in ic pos_trailer in
-    let magic_number = really_input_string ic len_magic_number in
-    if magic_number = Config.exec_magic_number then begin
-      dump_byte ic;
-      close_in ic
-    end else if Filename.check_suffix filename ".cmxs" then begin
-      flush stdout;
-      match read_dyn_header filename ic with
+    let expected_kind = Some Exec in
+    match read_current_info ~expected_kind ic with
+      | Error ((Unexpected_error _) as err) ->
+         exit_magic_error ~expected_kind err
+      | Ok _ ->
+         dump_obj_by_kind filename ic Exec;
+         Ok ()
+      | Error (Parse_error _)  ->
+         Error ()
+  and dump_cmxs ic =
+    flush stdout;
+    match find_dyn_offset filename with
       | None ->
-          printf "Unable to read info on file %s\n" filename;
-          exit 2
-      | Some header ->
-          if header.dynu_magic = Config.cmxs_magic_number then
-            print_cmxs_infos header
-          else begin
-            printf "Wrong magic number\n"; exit 2
-          end;
-          close_in ic
-    end else begin
-      printf "Not an OCaml object file\n"; exit 2
-    end
-  end
+         exit_errf "Unable to read info on %s %s."
+           (human_name_of_kind Cmxs) filename
+      | Some offset ->
+         LargeFile.seek_in ic offset;
+         let header = (input_value ic : dynheader) in
+         let expected_kind = Some Cmxs in
+         match parse header.dynu_magic with
+           | Error err ->
+              exit_magic_error ~expected_kind (Parse_error err)
+           | Ok info ->
+         match check_current Cmxs info with
+           | Error err ->
+              exit_magic_error ~expected_kind (Unexpected_error err)
+           | Ok () ->
+         LargeFile.seek_in ic offset;
+         dump_obj_by_kind filename ic Cmxs;
+         ()
+  in
+  printf "File %s\n" filename;
+  let ic = open_in_bin filename in
+  match dump_standard ic with
+    | Ok () -> ()
+    | Error head_error ->
+  match dump_exec ic with
+    | Ok () -> ()
+    | Error () ->
+  if Filename.check_suffix filename ".cmxs"
+  then dump_cmxs ic
+  else exit_magic_error ~expected_kind:None (Parse_error head_error)
```

Comment:
I understand your fondness for your newly coined indentation style but it's 1) not particularly standard 2) gives the wrong intuition about how things are nested 3) only reduce indentation depth by 2. Please use the regular style in the compiler. :)